src/main/java/picard/sam/markduplicates/UmiAwareDuplicateSetIterator.java

```diff
@@ -0,0 +1,225 @@
+import picard.PicardException;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Created by fleharty on 5/23/16.
+ */
+public class UmiAwareDuplicateSetIterator implements CloseableIterator<DuplicateSet> {
+    private DuplicateSetIterator wrappedIterator;
+    private Iterator<DuplicateSet> nextSetsIterator;
+    private List<List<Integer>> adjacencyList;
+    private List<Integer> groups;
+    int editDistanceToJoin;
+    boolean addInferredUmi;
+
+    public UmiAwareDuplicateSetIterator(final DuplicateSetIterator wrappedIterator, final int editDistanceToJoin, boolean addInferredUmi) {
+        this.wrappedIterator = wrappedIterator;
+        this.editDistanceToJoin = editDistanceToJoin;
+        this.addInferredUmi = addInferredUmi;
+        nextSetsIterator = Collections.emptyIterator();
+    }
+
+    @Override
+    public void close() {
+        wrappedIterator.close();
+    }
+
+    @Override
+    public boolean hasNext() {
+        return nextSetsIterator.hasNext() || wrappedIterator.hasNext();
+    }
+
+    @Override
+    public DuplicateSet next() {
+        if(!nextSetsIterator.hasNext())
+            process(wrappedIterator.next());
+
+        return nextSetsIterator.next();
+    }
+
+    // Takes a duplicate set and breaks it up into possible smaller sets according to the UMI,
+    // and updates nextSetsIterator to be an iterator on that set of DuplicateSets.
+    private void process(final DuplicateSet set) {
+
+        List<SAMRecord> records = set.getRecords();
+
+        groups = new ArrayList<>();
+        adjacencyList = new ArrayList<>();
+
+        // If any records don't have the RX attribute... don't do anything special
+        for(int i = 0;i < records.size();i++) {
+            if(records.get(i).getAttribute("RX") == null) {
+                nextSetsIterator = Collections.singleton(set).iterator();
+                return;
+            } else {
+            }
+        }
+
+        // Sort records by RX tag
+        Collections.sort(records, new Comparator<SAMRecord>() {
+            @Override
+            public int compare(final SAMRecord lhs, final SAMRecord rhs) {
+                if(lhs == null || rhs == null) {
+                    return 0;
+                } else if(lhs.getAttribute("RX") == null || rhs.getAttribute("RX") == null) {
+                    return 0;
+                }
+                else {
+                    return ((String) lhs.getAttribute("RX")).compareTo((String) rhs.getAttribute("RX"));
+                }
+            }
+        });
+
+        int n = records.size();
+
+        // Locate records that have identical UMI sequences
+        List<String> uniqueObservedUMIs = new ArrayList<String>();
+
+        uniqueObservedUMIs.add((String) records.get(0).getAttribute("RX"));
+        for(int i = 1;i < n;i++) {
```

Comment:
I cleaned this up in a few ways, but perhaps not in the Java 8 way... I also tried using a stream here, but it was even uglier.